ydb/core/cms/console/console_handshake.cpp

```diff
@@ -0,0 +1,144 @@
+        SendInReply(ev.Sender, ev.InterconnectSession, std::move(failEvent));
+        return false;
+    } else if (Self.CurrentPipeServerId != ev.Recipient) {
+        return false;
+    }
+    return true;
+}
+
+void TConfigsManager::Handle(TEvBlobStorage::TEvControllerProposeConfigRequest::TPtr &ev, const TActorContext &ctx) {
+    const auto& record = ev->Get()->Record;
+    const auto& proposedConfigHash = record.GetConfigHash();
+    const auto& proposedConfigVersion = record.GetConfigVersion();
+
+    ui64 currentConfigHash = NKikimr::NYaml::GetConfigHash(YamlConfig);
+    if (Self.CurrentSenderId != ev->Sender) {
+        auto endSessionEvent = std::make_unique<TEvents::TEvPoisonPill>();
+        ctx.Send(Self.CurrentPipeServerId, endSessionEvent.release());
+    }
+
+    Self.CurrentSenderId = ev->Sender;
+    Self.CurrentPipeServerId = ev->Recipient;
+    auto response = std::make_unique<TEvBlobStorage::TEvControllerProposeConfigResponse>();
+    auto& responseRecord = response->Record;
+
+    if (YamlVersion > record.GetConfigVersion()) {
+        responseRecord.SetStatus(NKikimrBlobStorage::TEvControllerProposeConfigResponse::UnexpectedConfig);
+        responseRecord.SetProposedConfigVersion(proposedConfigVersion);
+        responseRecord.SetConsoleConfigVersion(YamlVersion);
+        SendInReply(ev->Sender, ev->InterconnectSession, std::move(response));
+        LOG_ALERT_S(ctx, NKikimrServices::CMS, "Unexpected proposed config.");
+        return;
+    }
+    if (YamlVersion == record.GetConfigVersion()) {
+        if (proposedConfigHash != currentConfigHash) {
+            responseRecord.SetStatus(NKikimrBlobStorage::TEvControllerProposeConfigResponse::HashMismatch);
+            responseRecord.SetProposedConfigHash(proposedConfigHash);
+            responseRecord.SetConsoleConfigHash(currentConfigHash);
+            SendInReply(ev->Sender, ev->InterconnectSession, std::move(response));
+            LOG_ALERT_S(ctx, NKikimrServices::CMS, "Config hash mismatch.");
+            return;
+        }
+        responseRecord.SetStatus(NKikimrBlobStorage::TEvControllerProposeConfigResponse::CommitIsNotNeeded);
+        SendInReply(ev->Sender, ev->InterconnectSession, std::move(response));
+        return;
+    }
+    responseRecord.SetStatus(NKikimrBlobStorage::TEvControllerProposeConfigResponse::CommitIsNeeded);
+    SendInReply(ev->Sender, ev->InterconnectSession, std::move(response));
+}
+
+void TConfigsManager::Handle(TEvBlobStorage::TEvControllerConsoleCommitRequest::TPtr &ev, const TActorContext &ctx) {
+    auto response = std::make_unique<TEvBlobStorage::TEvControllerConsoleCommitResponse>();
+    const auto& yamlConfig = ev->Get()->Record.GetYAML();
+    if (!CheckSession(*ev, response, NKikimrBlobStorage::TEvControllerConsoleCommitResponse::SessionMismatch)) {
+        return;
+    }
+
+    auto *actor = new TConsoleCommitActor(Self, ev->Sender, yamlConfig, ev->InterconnectSession);
```

Comment:
yep, you're letting Self to get accessed in (1) another thread (2) and that actor may outlive this one